src/core/net/dns_dso.cpp

```diff
@@ -0,0 +1,1537 @@
+// otPlatDso transport callbacks
+
+extern "C" otInstance *otPlatDsoGetInstance(otPlatDsoConnection *aConnection)
+{
+    return &AsCoreType(aConnection).GetInstance();
+}
+
+extern "C" otPlatDsoConnection *otPlatDsoAccept(otInstance *aInstance, const otSockAddr *aPeerSockAddr)
+{
+    return AsCoreType(aInstance).Get<Dso>().AcceptConnection(AsCoreType(aPeerSockAddr));
+}
+
+extern "C" void otPlatDsoHandleConnected(otPlatDsoConnection *aConnection)
+{
+    AsCoreType(aConnection).HandleConnected();
+}
+
+extern "C" void otPlatDsoHandleReceive(otPlatDsoConnection *aConnection, otMessage *aMessage)
+{
+    AsCoreType(aConnection).HandleReceive(AsCoreType(aMessage));
+}
+
+extern "C" void otPlatDsoHandleDisconnected(otPlatDsoConnection *aConnection, otPlatDsoDisconnectMode aMode)
+{
+    AsCoreType(aConnection).HandleDisconnected(MapEnum(aMode));
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+// Dso::Connection
+
+Dso::Connection::Connection(Instance &           aInstance,
+                            const Ip6::SockAddr &aPeerSockAddr,
+                            Callbacks &          aCallbacks,
+                            uint32_t             aInactivityTimeout,
+                            uint32_t             aKeepAliveInterval)
+    : InstanceLocator(aInstance)
+    , mNext(nullptr)
+    , mCallbacks(aCallbacks)
+    , mPeerSockAddr(aPeerSockAddr)
+    , mState(kStateDisconnected)
+    , mIsServer(false)
+    , mInactivity(aInactivityTimeout)
+    , mKeepAlive(aKeepAliveInterval)
+{
+    OT_ASSERT(aKeepAliveInterval >= kMinKeepAliveInterval);
+    Init();
+}
+
+void Dso::Connection::Init(void)
+{
+    mNextMessageId       = 1;
+    mStateDidChange      = false;
+    mLongLivedOperation  = false;
+    mRetryDelay          = 0;
+    mRetryDelayErrorCode = Dns::Header::kResponseSuccess;
+    mDisconnectReason    = kReasonUnknown;
+}
+
+void Dso::Connection::SetState(State aState)
+{
+    VerifyOrExit(mState != aState);
+
+    otLogInfoDns("[dso] State: %s -> %s on connection with %s", StateToString(mState), StateToString(aState),
+                 mPeerSockAddr.ToString().AsCString());
+
+    mState          = aState;
+    mStateDidChange = true;
+
+exit:
+    return;
+}
+
+void Dso::Connection::SignalAnyStateChange(void)
+{
+    VerifyOrExit(mStateDidChange);
+    mStateDidChange = false;
+
+    switch (mState)
+    {
+    case kStateDisconnected:
+        mCallbacks.mHandleDisconnected(*this);
+        break;
+
+    case kStateConnectedButSessionless:
+        mCallbacks.mHandleConnected(*this);
+        break;
+
+    case kStateSessionEstablished:
+        mCallbacks.mHandleSessionEstablished(*this);
+        break;
+
+    case kStateConnecting:
+    case kStateEstablishingSession:
+        break;
+    };
+
+exit:
+    return;
+}
+
+Message *Dso::Connection::NewMessage(void)
```

Comment:
We have `otIp6NewMessage()`

Reply:
I figure otIp6NewMessage() can create a buffer. 

However, a TCP DNS message can easily exceed the capacity of one message buffer?
Instead of receiving all data of a DNS message, which may require significant buffer size, we may better need to create a message buffer when the length is received, and continue to append new data to the otMessage when more data is received?

Do we have APIs for constructing a large otMessage to hold incoming data stream?

Reply:
`otMessage()` does not technically have a size limit. It automatically grows as new data is appended/written to it (e.g. using `otMessageAppend()`).
-  The underlying `Message` implementation will allocate and chain new buffer blocks as new data is added, but that is not visible to the users of `otMessage`. 
-  I guess the limit is technically the overall buffer pool size, but a single `otMessage` can potentially use all buffers from pool and entire pool size, so it can grow very large (also message length is `uint16_t` which limit the length to  ~65K).